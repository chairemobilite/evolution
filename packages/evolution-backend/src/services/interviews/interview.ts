/*
 * Copyright 2022, Polytechnique Montreal and contributors
 *
 * This file is licensed under the MIT License.
 * License text available at https://opensource.org/licenses/MIT
 */
import _set from 'lodash.set';
import _unset from 'lodash.unset';
import moment from 'moment';
import { UserAttributes } from 'chaire-lib-backend/lib/services/auth/user';
import serverValidate, { ServerValidation } from '../validations/serverValidation';
import serverUpdateField, { ServerFieldUpdateCallback } from './serverFieldUpdate';
import config from 'chaire-lib-backend/lib/config/server.config';
import interviewsDbQueries from '../../models/interviews.db.queries';
import { UserInterviewAttributes, InterviewAttributes } from 'evolution-common/lib/services/interviews/interview';
import projectConfig from '../../config/projectConfig';

export const addRolesToInterview = <CustomSurvey, CustomHousehold, CustomHome, CustomPerson>(
    interview: UserInterviewAttributes<CustomSurvey, CustomHousehold, CustomHome, CustomPerson>,
    user: UserAttributes
) => {
    // Add the userRoles in the interview object
    const permissions = user.permissions;
    interview.userRoles = permissions ? Object.keys(permissions).filter((perm) => permissions[perm] === true) : [];
};

export const setInterviewFields = <CustomSurvey, CustomHousehold, CustomHome, CustomPerson>(
    interview: InterviewAttributes<CustomSurvey, CustomHousehold, CustomHome, CustomPerson>,
    options: {
        valuesByPath: { [key: string]: unknown };
        unsetPaths?: string[];
    }
) => {
    // Set the interview's values by path received from client
    for (const path in options.valuesByPath) {
        const value = options.valuesByPath[path];
        _set(interview, path, value);
    }
    // Unset any interview path defined by client
    if (Array.isArray(options.unsetPaths)) {
        for (let i = 0, count = options.unsetPaths.length; i < count; i++) {
            _unset(interview, options.unsetPaths[i]);
        }
    }
};

/**
 * Update an interview, given the paths to update. If database logs are
 * configured, a new log will be saved for this update.
 *
 * TODO: Rename to save, as it conveys more the fact that it is saved in the db than update
 *
 * @param {InterviewAttributes} interview The base interview to update, as
 * fetched in the database
 * @param {{valuesByPath: { [key: string]: unknown }; unsetPaths?: string[];
 * serverValidations?: ServerValidation; fieldsToUpdate?: (keyof
 * InterviewAttributes)[]; logData?: { [key: string]: unknown };}} options The
 * `valuesByPath` is a key-value map of values to update in the interview. The
 * key is the dot-separated hierarchical path to the field to update. For
 * example, to update a field in response, it could be `{
 * 'responses.mySection.myField': 'myValue' }`. The `unsetPaths` array is an
 * array of dot-separated paths whose value to unset. The 'serverValidations'
 * are the validations to do on the data. If any new field to set does not
 * validate, the value will be set in the responses object, but a corresponding
 * value will be set to false in the validations object. The 'fieldsToUpdate'
 * makes sure only those fields are updated in the interview. Since there is no
 * control on the paths of the valuesByPath keys, it avoids updating fields in
 * code path where they should not be updated. The 'logData' object contains
 * data that should be added to the log object of the current update, if log is
 * enabled.
 **/
export const updateInterview = async <CustomSurvey, CustomHousehold, CustomHome, CustomPerson>(
    interview: InterviewAttributes<CustomSurvey, CustomHousehold, CustomHome, CustomPerson>,
    options: {
        logDatabaseUpdates?: boolean;
        valuesByPath: { [key: string]: unknown };
        unsetPaths?: string[];
        serverValidations?: ServerValidation;
        fieldsToUpdate?: (keyof InterviewAttributes<CustomSurvey, CustomHousehold, CustomHome, CustomPerson>)[];
        logData?: { [key: string]: unknown };
    }
): Promise<{
    interviewId: string | undefined;
    serverValidations:
        | true
        | {
              [key: string]: {
                  [key: string]: string;
              };
          };
    serverValuesByPath: { [key: string]: unknown };
}> => {
    const fieldsToUpdate = options.fieldsToUpdate || ['responses', 'validations'];
    const logData = options.logData || {};
    const serverValidations = await serverValidate(
        options.serverValidations,
        options.valuesByPath,
        options.unsetPaths || []
    );
    // Update values by path with caller provided values
    setInterviewFields(interview, { valuesByPath: options.valuesByPath, unsetPaths: options.unsetPaths });
    const serverValuesByPath = await serverUpdateField(
        interview,
        projectConfig.serverUpdateCallbacks,
        options.valuesByPath,
        options.unsetPaths
    );
    // Update values with server updated values
    if (Object.keys(serverValuesByPath).length > 0) {
        setInterviewFields(interview, { valuesByPath: serverValuesByPath, unsetPaths: options.unsetPaths });
    }
    // If server validation failed, add the invalid fields to validation
    if (serverValidations !== true) {
        for (const path in serverValidations) {
            _set(interview.validations, `${path}`, false);
        }
    }
    interview.responses = interview.responses || {};
    interview.validations = interview.validations || {};
    if (!interview.logs || !Array.isArray(interview.logs)) {
        interview.logs = [];
        //console.log(interview.logs);
    }

    const databaseUpdateJson: Partial<InterviewAttributes<CustomSurvey, CustomHousehold, CustomHome, CustomPerson>> =
        {};
    fieldsToUpdate.forEach((field) => {
        // FIXME: For some reason, the following line gives type error, not sure why, so casting to any and disabling the warning.
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        (databaseUpdateJson as any)[field] = interview[field];
    });

    // logs if configured:
    // FIXME Type evolution's configs
    if (options.logDatabaseUpdates !== false && (config as any).logDatabaseUpdates) {
        const timestamp = interview.responses._updatedAt || moment().unix();
        if (options.unsetPaths) {
            interview.logs.push({
                ...logData,
                timestamp,
                valuesByPath: options.valuesByPath,
                unsetPaths: options.unsetPaths
            });
        } else {
            interview.logs.push({
                ...logData,
                timestamp,
                valuesByPath: options.valuesByPath
            });
        }
        databaseUpdateJson.logs = interview.logs;
    }

    const retInterview = await interviewsDbQueries.update(interview.uuid, databaseUpdateJson);
    return { interviewId: retInterview.uuid, serverValidations, serverValuesByPath };
};
