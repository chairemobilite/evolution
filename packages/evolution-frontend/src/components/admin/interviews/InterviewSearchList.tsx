/*
 * Copyright 2022, Polytechnique Montreal and contributors
 *
 * This file is licensed under the MIT License.
 * License text available at https://opensource.org/licenses/MIT
 */
import React from 'react';
import { WithTranslation, withTranslation } from 'react-i18next';

import Interview from './InterviewSearchResult';
import { InterviewContext } from '../../../contexts/InterviewContext';
import { _isBlank } from 'chaire-lib-common/lib/utils/LodashExtensions';
import InterviewsCreateLink from './InterviewsCreateLink';

type InterviewSearchListProps = {
    autoCreateIfNoData: boolean;
    queryData: URLSearchParams;
};

const InterviewSearchList: React.FunctionComponent<InterviewSearchListProps & WithTranslation> = (
    props: InterviewSearchListProps & WithTranslation
) => {
    const [data, setData] = React.useState<{ [key: string]: any }[]>([]);
    const [loading, setLoading] = React.useState(false);
    const { state, dispatch } = React.useContext(InterviewContext);
    const fetchIdRef = React.useRef(0);

    // Function to fetch data from the server, with paging and filtering
    const fetchData = React.useCallback(
        async (accessCode: string) => {
            // Give this fetch an ID
            const fetchId = ++fetchIdRef.current;

            // Set the loading state
            setLoading(true);

            // Make a query string from the filters

            try {
                const response = await fetch(`/api/interviews/interviewByCode?accessCode=${accessCode}`);
                if (fetchId !== fetchIdRef.current) {
                    // There was another query since, ignore
                    return;
                }
                if (response.status === 200) {
                    const jsonData = await response.json();
                    if (jsonData.interviews) {
                        setData(jsonData.interviews);
                        if (
                            !_isBlank(accessCode) &&
                            jsonData.interviews.length === 0 &&
                            props.autoCreateIfNoData &&
                            state.status !== 'creating'
                        ) {
                            // automatically create new interview with this access code
                            dispatch({
                                type: 'createNew',
                                username: `telephone_${(Math.ceil(Math.random() * 8999) + 1000).toString()}`,
                                queryData: props.queryData
                            });
                        }
                    }
                } else {
                    console.error('Invalid response code from server: ', response.status);
                }
            } catch (error) {
                console.error(`Error fetching user data from server: ${error}`);
                setData([]);
            } finally {
                if (fetchId === fetchIdRef.current) {
                    setLoading(false);
                }
            }
        },
        [state.status]
    );

    React.useEffect(() => {
        fetchData(state.responses.accessCode);
    }, [state.responses]);

    // TODO This is very specific to the interviewer use case, generalize it and use a function to generate username, maybe as prop
    return (
        <ul className="interview">
            {data.map((interview) => (
                <Interview key={interview.uuid} interview={interview} />
            ))}
            <InterviewsCreateLink queryData={props.queryData} />
        </ul>
    );
};

export default withTranslation(['admin', 'main'])(InterviewSearchList);
