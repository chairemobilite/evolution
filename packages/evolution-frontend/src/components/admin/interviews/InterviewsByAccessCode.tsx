/*
 * Copyright 2022, Polytechnique Montreal and contributors
 *
 * This file is licensed under the MIT License.
 * License text available at https://opensource.org/licenses/MIT
 */
import React from 'react';
import { useTranslation } from 'react-i18next';
import Loadable from 'react-loadable';
import Loader from 'react-spinners/HashLoader';

import { InterviewContext } from '../../../contexts/InterviewContext';
import InputString from 'chaire-lib-frontend/lib/components/input/InputString';
import { Link, useLocation, useParams } from 'react-router';
import { _booleish } from 'chaire-lib-common/lib/utils/LodashExtensions';

const loader = function Loading() {
    return <Loader size={30} color={'#aaaaaa'} loading={true} />;
};

const InterviewsComponent = Loadable({
    // TODO: move this to componentDidMount like in Monitoring.tsx
    loader: () => import('./InterviewSearchList'),
    loading: loader
});

const InterviewsByAccessCode: React.FC = () => {
    const location = useLocation();
    const params = useParams();
    const { t } = useTranslation('admin');

    const urlSearch = new URLSearchParams(location.search);
    const [currentCode, setCurrentCode] = React.useState(params.accessCode);
    const [createNewIfNoData] = React.useState(_booleish(urlSearch.get('autoCreate')) === true);
    const { state, dispatch } = React.useContext(InterviewContext);

    React.useEffect(() => {
        if (params.accessCode !== state.response.accessCode) {
            dispatch({ type: 'update_response', response: { accessCode: params.accessCode } });
        }
    }, [params.accessCode, state.response]);

    return (
        <div className="admin">
            <div className="survey-section__content apptr__form-container">
                <InputString
                    id="accessCodeSearchInput"
                    value={currentCode}
                    onValueUpdated={(newValue) => setCurrentCode(newValue.value)}
                ></InputString>
                <Link to={`/interviews/byCode/${currentCode}`}>
                    <button type="button" className={'survey-section__button button blue small'}>
                        {t('admin:interviewSearch:SearchByCode')}
                    </button>
                </Link>
            </div>
            <InterviewsComponent autoCreateIfNoData={createNewIfNoData} queryData={urlSearch} />
        </div>
    );
};

export default InterviewsByAccessCode;
