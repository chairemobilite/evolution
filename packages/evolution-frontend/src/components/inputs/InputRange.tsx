/*
 * Copyright 2022, Polytechnique Montreal and contributors
 *
 * This file is licensed under the MIT License.
 * License text available at https://opensource.org/licenses/MIT
 */
import React from 'react';
// TODO this package hasn't been updated in a while. Consider changing for a more maintained one
import ReactInputRange, { Range } from 'react-input-range';
import { withTranslation, WithTranslation } from 'react-i18next';
import defaultClassNames from 'react-input-range/src/js/input-range/default-class-names';
import 'react-input-range/lib/css/index.css';

import { _isBlank } from 'chaire-lib-common/lib/utils/LodashExtensions';
import { InputRangeType } from 'evolution-common/lib/services/widgets';
import * as surveyHelper from 'evolution-common/lib/utils/helpers';
import { CommonInputProps } from './CommonInputProps';

export type InputRangeProps<CustomSurvey, CustomHousehold, CustomHome, CustomPerson> = CommonInputProps<
    CustomSurvey,
    CustomHousehold,
    CustomHome,
    CustomPerson
> & {
    value?: number;
    inputRef?: React.LegacyRef<ReactInputRange>;
    widgetConfig: InputRangeType<CustomSurvey, CustomHousehold, CustomHome, CustomPerson>;
};

export const InputRange = <CustomSurvey, CustomHousehold, CustomHome, CustomPerson>(
    props: InputRangeProps<CustomSurvey, CustomHousehold, CustomHome, CustomPerson> & WithTranslation
) => {
    const [value, setValue] = React.useState<number | Range | undefined>(props.value);

    const labels = props.widgetConfig.labels;
    const labelItems = labels
        ? labels.map((label, idx) => {
            const labelStr = surveyHelper.translateString(label, props.i18n, props.interview, props.path, props.user);
            return (
                <p
                    key={`inputRange_${props.id}_label_${idx}`}
                    className={`__input-range_input_container_label${
                        idx > 0 && idx < labels.length - 1
                            ? ' __input-range_input_container_label_middle'
                            : idx === labels.length - 1
                                ? ' __input-range_input_container_label_last'
                                : ''
                    }`}
                >
                    {labelStr}
                </p>
            );
        })
        : [];

    if (props.widgetConfig.trackClassName) {
        defaultClassNames.track = props.widgetConfig.trackClassName;
    }

    const formatLabelFct = props.widgetConfig.formatLabel;

    const maxValue: number = !_isBlank(props.widgetConfig.maxValue) ? props.widgetConfig.maxValue || 100 : 100;
    const minValue: number = !_isBlank(props.widgetConfig.minValue) ? props.widgetConfig.minValue || -10 : -10;

    return (
        <div
            className={`survey-question__input-range_input_container_with_labels${
                _isBlank(value) ? ' input-blank' : ''
            }`}
        >
            <ReactInputRange
                name={props.id}
                // defaultValue = {_isBlank(this.props.value) ? this.props.widgetConfig.defaultValue : this.props.value }
                ref={props.inputRef}
                maxValue={maxValue as number | undefined}
                minValue={minValue as number | undefined}
                formatLabel={
                    formatLabelFct !== undefined
                        ? (value) => formatLabelFct(value, props.i18n.language)
                        : (value) => String(value)
                }
                value={!_isBlank(value) && value !== undefined && value >= minValue ? value : minValue}
                classNames={defaultClassNames}
                onChange={(value) => setValue(value)}
                onChangeComplete={(value) => {
                    if (value < minValue) {
                        props.onValueChange({ target: { value: null } });
                    } else {
                        props.onValueChange({ target: { value } });
                    }
                }}
            />
            {labelItems.length > 0 && (
                <div className={'survey-question__input-range_input_labels_container'}>{labelItems}</div>
            )}
        </div>
    );
};

export default withTranslation()(InputRange) as React.FunctionComponent<InputRangeProps<any, any, any, any>>;
