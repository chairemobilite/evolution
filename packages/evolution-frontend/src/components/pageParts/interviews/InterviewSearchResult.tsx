/*
 * Copyright 2022, Polytechnique Montreal and contributors
 *
 * This file is licensed under the MIT License.
 * License text available at https://opensource.org/licenses/MIT
 */
import React from 'react';
import { WithTranslation, withTranslation } from 'react-i18next';

interface InterviewProps extends WithTranslation {
    interview: { [key: string]: any };
    key: string;
}

const InterviewSearchResult: React.FunctionComponent<InterviewProps> = (props: InterviewProps) => {
    const interviewUrl = `/survey/edit/${props.interview.uuid}/`;
    return (
        <li
            title={props.interview.uuid}
            key={props.key}
            className={`
                ${props.interview.isValid === true && props.interview.isCompleted === true ? '_dark-green _strong' : ''}
                ${
        props.interview.isValidated === true &&
                    props.interview.isValid === true &&
                    props.interview.isCompleted === true
            ? '_green _strong _active-background'
            : ''
        }
                ${props.interview.isValid === true && props.interview.isCompleted === false ? '_orange _strong' : ''}
                ${props.interview.isValid === false ? '_dark-red _strong' : ''}
                `}
        >
            {props.t('admin:interviewSearch:InterviewUser')}
            {': '}
            {props.interview.email ||
                props.interview.username ||
                (props.interview.facebook
                    ? props.t('admin:interviewSearch:FromFacebook')
                    : props.interview.google
                        ? props.t('admin:interviewSearch:FromGoogle')
                        : props.t('admin:interviewSearch:UnknownUser'))}{' '}
            •{' '}
            {props.interview.isCompleted === true
                ? props.t('admin:CompletedFemSingular')
                : props.interview.isCompleted === false
                    ? props.t('admin:NotCompletedFemSingular')
                    : props.t('admin:UnknownCompletionFemSingular')}{' '}
            •{' '}
            {props.interview.isValid === true
                ? props.t('admin:Valid')
                : props.interview.isValid === false
                    ? props.t('admin:Invalid')
                    : props.t('admin:UnknownValidity')}{' '}
            • {props.interview.home.address || props.t('admin:interviewSearch:UnknownAddress')},{' '}
            {props.interview.home.city || ''} •{' '}
            <a
                href={interviewUrl}
                id={`interviewButtonList_${props.interview.uuid}`}
                data-uuid={props.interview.uuid}
                target="_blank"
                rel="noreferrer"
            >
                {props.t('admin:interviewSearch:Open')}
            </a>
        </li>
    );
};

export default withTranslation(['admin', 'main'])(InterviewSearchResult);
