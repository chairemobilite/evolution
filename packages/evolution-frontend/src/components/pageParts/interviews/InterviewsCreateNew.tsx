/*
 * Copyright 2022, Polytechnique Montreal and contributors
 *
 * This file is licensed under the MIT License.
 * License text available at https://opensource.org/licenses/MIT
 */
import React from 'react';
import { createBrowserHistory } from 'history';
import Loader from 'react-spinners/HashLoader';
import { InterviewContext } from '../../../contexts/InterviewContext';

const InterviewsCreateNew: React.FunctionComponent = () => {
    const [inProgress, setInProgress] = React.useState(false);
    const { state, dispatch } = React.useContext(InterviewContext);

    // Function to fetch data from the server, with paging and filtering
    const createUser = React.useCallback(async () => {
        if (inProgress || state.status !== 'creating') {
            return;
        }

        // Set in progress state
        setInProgress(true);

        // Make a query string from the filters
        try {
            const response = await fetch('/api/interviews/createNew', {
                headers: {
                    Accept: 'application/json',
                    'Content-Type': 'application/json'
                },
                credentials: 'include',
                method: 'POST',
                body: JSON.stringify({
                    createUser: state.username,
                    responses: state.responses
                })
            });
            const history = createBrowserHistory({ forceRefresh: true });
            if (response.status === 200) {
                const jsonData = await response.json();
                if (jsonData.interviewUuid) {
                    history.push(`/survey/edit/${jsonData.interviewUuid}/`);
                    dispatch({ type: 'success', interviewUuid: jsonData.interviewUuid });
                    return;
                }
                console.log('Server did not return a proper interview id', jsonData);
                // Goto proper error page
            }
            // TODO Go to proper status code page
            console.log('Bad response code from server', response.status);
        } catch (error) {
            // TODO Go to proper error page
            console.error(`Error fetching user data from server: ${error}`);
        }

        // Cancel in progress state, only if the page was not redirected
        setInProgress(false);
    }, []);

    React.useEffect(() => {
        createUser();
    }, []);

    return (
        <div className="admin">
            {inProgress && <Loader size={30} color={'#aaaaaa'} loading={true} />}
            {!inProgress && 'error'}
        </div>
    );
};

/** Component used to create new interviews by users other than the main user */
export default InterviewsCreateNew;
