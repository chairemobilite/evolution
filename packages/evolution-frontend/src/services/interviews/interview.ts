import { UserInterviewAttributes } from 'evolution-common/lib/services/interviews/interview';
import { WidgetConfig, LangData } from '../widgets';

export type WidgetStatus<CustomSurvey, CustomHousehold, CustomHome, CustomPerson> = {
    path: string;
    customPath?: string;
    isVisible: boolean;
    modalIsOpen: boolean;
    isDisabled: boolean;
    isCollapsed: boolean;
    isEmpty: boolean;
    isCustomEmpty: boolean;
    isValid: boolean;
    isResponded: boolean;
    isCustomResponded: boolean;
    errorMessage?: LangData<CustomSurvey, CustomHousehold, CustomHome, CustomPerson>;
    groupedObjectId?: string;
    value: unknown;
    customValue?: unknown;
    /** Key that should be changed if the widget is updated outside the application workflow, eg. from the server */
    currentUpdateKey: number;
};

/* FIXME This type represents what was in the legacy evolution, with fields that
 evolve throughout the application, they represent state of the local interview
 more than interview attributes. This all should be moved to some class, with
 methods to follow the state of the current interview. */
export type FrontendInterviewAttributes<CustomSurvey, CustomHousehold, CustomHome, CustomPerson> = {
    previousWidgets?: {
        [widgetShortname: string]: WidgetStatus<CustomSurvey, CustomHousehold, CustomHome, CustomPerson>;
    };
    previousGroups?: {
        [groupName: string]: {
            [groupId: string]: {
                [widgetShortname: string]: WidgetStatus<CustomSurvey, CustomHousehold, CustomHome, CustomPerson>;
            };
        };
    };
    // These are widgets statuses for the current section, if they are not grouped
    widgets: { [widgetShortname: string]: WidgetStatus<CustomSurvey, CustomHousehold, CustomHome, CustomPerson> };
    // These are the widget status for the groups in the current section
    groups: {
        [groupName: string]: {
            [groupId: string]: {
                [widgetShortname: string]: WidgetStatus<CustomSurvey, CustomHousehold, CustomHome, CustomPerson>;
            };
        };
    };
    // Contains the paths in the responses of the visible widgets... FIXME Rename?
    visibleWidgets: string[];
    allWidgetsValid: boolean;
    // Name of the currently loaded section
    sectionLoaded?: string;
};

export type UserFrontendInterviewAttributes<CustomSurvey, CustomHousehold, CustomHome, CustomPerson> =
    FrontendInterviewAttributes<CustomSurvey, CustomHousehold, CustomHome, CustomPerson> &
        UserInterviewAttributes<CustomSurvey, CustomHousehold, CustomHome, CustomPerson>;

export type SurveySectionGroup = {
    widgets: string[];
};

// TODO Properly type this
export type SurveySection = {
    widgets: string[];
    groups?: {
        [groupName: string]: SurveySectionGroup;
    };
    [key: string]: unknown;
};
export type SurveySections = { [sectionName: string]: SurveySection };
export type SurveyWidgets<CustomSurvey, CustomHousehold, CustomHome, CustomPerson> = {
    [widgetName: string]: WidgetConfig<CustomSurvey, CustomHousehold, CustomHome, CustomPerson>;
};
