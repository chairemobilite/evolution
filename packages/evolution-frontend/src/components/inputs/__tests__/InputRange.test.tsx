/*
 * Copyright 2022, Polytechnique Montreal and contributors
 *
 * This file is licensed under the MIT License.
 * License text available at https://opensource.org/licenses/MIT
 */
import { UserInterviewAttributes } from 'evolution-common/lib/services/interviews/interview';
import React from 'react';
import TestRenderer from 'react-test-renderer';

import InputRange from '../InputRange';

// Mock the react-input-range files to avoid jest compilation errors
jest.mock('react-input-range/src/js/input-range/default-class-names', () => ({
    activeTrack: 'input-range__track input-range__track--active',
    disabledInputRange: 'input-range input-range--disabled',
    inputRange: 'input-range',
    labelContainer: 'input-range__label-container',
    maxLabel: 'input-range__label input-range__label--max',
    minLabel: 'input-range__label input-range__label--min',
    slider: 'input-range__slider',
    sliderContainer: 'input-range__slider-container',
    track: 'input-range__track input-range__track--background',
    valueLabel: 'input-range__label input-range__label--value',
}));

jest.mock('react-input-range/lib/css/index.css', () => {});

const userAttributes = {
    id: 1,
    username: 'foo',
    preferences: {  },
    serializedPermissions: [],
    isAuthorized: () => true,
    is_admin: false,
    pages: [],
    showUserInfo: true
}

const interviewAttributes: UserInterviewAttributes = {
    id: 1,
    uuid: 'arbitrary uuid',
    user_id: 1,
    is_completed: false,
    responses: {
        section1: {
            q1: 'abc',
            q2: 3
        },
        section2: {
            q1: 'test'
        }
    },
    validations: {
        section1: {
            q1: true,
            q2: false
        },
        section2: {
            q1: true
        }
    },
    is_valid: true
};

describe('Should correctly render InputRange with minimal parameters', () => {

    const widgetConfig = {
        type: 'question' as const,
        twoColumns: true,
        path: 'test.foo',
        containsHtml: true,
        label: {
            fr: `Texte en français`,
            en: `English text`
        },
        inputType: 'slider' as const
    }

    test('Test without value', () => {
        // Should have a blank style
        const wrapper = TestRenderer.create(
            <InputRange
                id={'test'}
                onValueChange={() => { /* nothing to do */}}
                widgetConfig={widgetConfig}
                value={undefined}
                inputRef={React.createRef()}
                interview={interviewAttributes}
                user={userAttributes}
                path='foo.test'
            />
        );
        expect(wrapper).toMatchSnapshot();
    });

    test('Test with value', () => {
        const wrapper = TestRenderer.create(
            <InputRange
                id={'test'}
                onValueChange={() => { /* nothing to do */}}
                widgetConfig={widgetConfig}
                value={10}
                inputRef={React.createRef()}
                interview={interviewAttributes}
                user={userAttributes}
                path='foo.test'
            />
        );
        expect(wrapper).toMatchSnapshot();
    });

});

describe('Should correctly render InputRange with various parameters', () => {

    const baseWidgetConfig = {
        type: 'question' as const,
        twoColumns: true,
        path: 'test.foo',
        containsHtml: true,
        label: {
            fr: `Texte en français`,
            en: `English text`
        },
        inputType: 'slider' as const
    }

    test('Test with min max range values', () => {
        const widgetConfig = Object.assign({ 
            maxValue: 40,
            minValue: 10
        }, baseWidgetConfig);
        const wrapper = TestRenderer.create(
            <InputRange
                id={'test'}
                onValueChange={() => { /* nothing to do */}}
                widgetConfig={widgetConfig}
                value={5}
                inputRef={React.createRef()}
                interview={interviewAttributes}
                user={userAttributes}
                path='foo.test'
            />
        );
        expect(wrapper).toMatchSnapshot();
    });

    test('Test trackclass and labels', () => {
        const formatLabel = jest.fn().mockImplementation((val) => `l${val}`);
        const widgetConfig = Object.assign({ 
            formatLabel,
            labels: ['unilingual text', { fr: '2e label français', en: '2nd english label' }],
            trackClassName: 'myTrackClass',
        }, baseWidgetConfig);
        const wrapper = TestRenderer.create(
            <InputRange
                id={'test'}
                onValueChange={() => { /* nothing to do */}}
                widgetConfig={widgetConfig}
                value={10}
                inputRef={React.createRef()}
                interview={interviewAttributes}
                user={userAttributes}
                path='foo.test'
            />
        );
        expect(wrapper).toMatchSnapshot();
    });

});
