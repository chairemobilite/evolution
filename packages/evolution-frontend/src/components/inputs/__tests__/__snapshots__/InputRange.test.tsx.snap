// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should correctly render InputRange with minimal parameters Test with value 1`] = `
<div
  className="survey-question__input-range_input_container_with_labels"
>
  <div
    aria-disabled={false}
    className="input-range"
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onTouchStart={[Function]}
  >
    <span
      className="input-range__label input-range__label--min"
    >
      <span
        className="input-range__label-container"
      >
        -10
      </span>
    </span>
    <div
      className="input-range__track input-range__track--background"
      onMouseDown={[Function]}
      onTouchStart={[Function]}
    >
      <div
        className="input-range__track input-range__track--active"
        style={
          Object {
            "left": "0%",
            "width": "18.181818181818183%",
          }
        }
      />
      <span
        className="input-range__slider-container"
        style={
          Object {
            "left": "18.181818181818183%",
            "position": "absolute",
          }
        }
      >
        <span
          className="input-range__label input-range__label--value"
        >
          <span
            className="input-range__label-container"
          >
            10
          </span>
        </span>
        <div
          aria-valuemax={100}
          aria-valuemin={-10}
          aria-valuenow={10}
          className="input-range__slider"
          draggable="false"
          onKeyDown={[Function]}
          onMouseDown={[Function]}
          onTouchStart={[Function]}
          role="slider"
          tabIndex="0"
        />
      </span>
    </div>
    <span
      className="input-range__label input-range__label--max"
    >
      <span
        className="input-range__label-container"
      >
        100
      </span>
    </span>
    <input
      name="test"
      type="hidden"
      value={10}
    />
  </div>
</div>
`;

exports[`Should correctly render InputRange with minimal parameters Test without value 1`] = `
<div
  className="survey-question__input-range_input_container_with_labels input-blank"
>
  <div
    aria-disabled={false}
    className="input-range"
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onTouchStart={[Function]}
  >
    <span
      className="input-range__label input-range__label--min"
    >
      <span
        className="input-range__label-container"
      >
        -10
      </span>
    </span>
    <div
      className="input-range__track input-range__track--background"
      onMouseDown={[Function]}
      onTouchStart={[Function]}
    >
      <div
        className="input-range__track input-range__track--active"
        style={
          Object {
            "left": "0%",
            "width": "0%",
          }
        }
      />
      <span
        className="input-range__slider-container"
        style={
          Object {
            "left": "0%",
            "position": "absolute",
          }
        }
      >
        <span
          className="input-range__label input-range__label--value"
        >
          <span
            className="input-range__label-container"
          >
            -10
          </span>
        </span>
        <div
          aria-valuemax={100}
          aria-valuemin={-10}
          aria-valuenow={-10}
          className="input-range__slider"
          draggable="false"
          onKeyDown={[Function]}
          onMouseDown={[Function]}
          onTouchStart={[Function]}
          role="slider"
          tabIndex="0"
        />
      </span>
    </div>
    <span
      className="input-range__label input-range__label--max"
    >
      <span
        className="input-range__label-container"
      >
        100
      </span>
    </span>
    <input
      name="test"
      type="hidden"
      value={-10}
    />
  </div>
</div>
`;

exports[`Should correctly render InputRange with various parameters Test trackclass and labels 1`] = `
<div
  className="survey-question__input-range_input_container_with_labels"
>
  <div
    aria-disabled={false}
    className="input-range"
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onTouchStart={[Function]}
  >
    <span
      className="input-range__label input-range__label--min"
    >
      <span
        className="input-range__label-container"
      >
        l-10
      </span>
    </span>
    <div
      className="myTrackClass"
      onMouseDown={[Function]}
      onTouchStart={[Function]}
    >
      <div
        className="input-range__track input-range__track--active"
        style={
          Object {
            "left": "0%",
            "width": "18.181818181818183%",
          }
        }
      />
      <span
        className="input-range__slider-container"
        style={
          Object {
            "left": "18.181818181818183%",
            "position": "absolute",
          }
        }
      >
        <span
          className="input-range__label input-range__label--value"
        >
          <span
            className="input-range__label-container"
          >
            l10
          </span>
        </span>
        <div
          aria-valuemax={100}
          aria-valuemin={-10}
          aria-valuenow={10}
          className="input-range__slider"
          draggable="false"
          onKeyDown={[Function]}
          onMouseDown={[Function]}
          onTouchStart={[Function]}
          role="slider"
          tabIndex="0"
        />
      </span>
    </div>
    <span
      className="input-range__label input-range__label--max"
    >
      <span
        className="input-range__label-container"
      >
        l100
      </span>
    </span>
    <input
      name="test"
      type="hidden"
      value={10}
    />
  </div>
  <div
    className="survey-question__input-range_input_labels_container"
  >
    <p
      className="__input-range_input_container_label"
    >
      unilingual text
    </p>
    <p
      className="__input-range_input_container_label __input-range_input_container_label_last"
    >
      2nd english label
    </p>
  </div>
</div>
`;

exports[`Should correctly render InputRange with various parameters Test with min max range values 1`] = `
<div
  className="survey-question__input-range_input_container_with_labels"
>
  <div
    aria-disabled={false}
    className="input-range"
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onTouchStart={[Function]}
  >
    <span
      className="input-range__label input-range__label--min"
    >
      <span
        className="input-range__label-container"
      >
        10
      </span>
    </span>
    <div
      className="input-range__track input-range__track--background"
      onMouseDown={[Function]}
      onTouchStart={[Function]}
    >
      <div
        className="input-range__track input-range__track--active"
        style={
          Object {
            "left": "0%",
            "width": "0%",
          }
        }
      />
      <span
        className="input-range__slider-container"
        style={
          Object {
            "left": "0%",
            "position": "absolute",
          }
        }
      >
        <span
          className="input-range__label input-range__label--value"
        >
          <span
            className="input-range__label-container"
          >
            10
          </span>
        </span>
        <div
          aria-valuemax={40}
          aria-valuemin={10}
          aria-valuenow={10}
          className="input-range__slider"
          draggable="false"
          onKeyDown={[Function]}
          onMouseDown={[Function]}
          onTouchStart={[Function]}
          role="slider"
          tabIndex="0"
        />
      </span>
    </div>
    <span
      className="input-range__label input-range__label--max"
    >
      <span
        className="input-range__label-container"
      >
        40
      </span>
    </span>
    <input
      name="test"
      type="hidden"
      value={10}
    />
  </div>
</div>
`;
